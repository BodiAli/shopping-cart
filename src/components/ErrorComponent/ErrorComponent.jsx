import { useRouteError } from "react-router-dom";
import styles from "./ErrorComponent.module.css";

function ErrorComponent() {
  const error = useRouteError();

  function handleClick() {
    window.location.reload();
  }

  return (
    <section className={styles.section}>
      <h2>Unexpected error occurred!</h2>
      {error.data && <p>{error.data}</p>}
      <p>{error.statusText || error.message}</p>
      <button onClick={handleClick}>Retry</button>
    </section>
  );
}
export default ErrorComponent;
