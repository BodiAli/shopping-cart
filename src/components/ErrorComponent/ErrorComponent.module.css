.section {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
  padding: 5rem;
  background-color: var(--green);
  min-height: 90vh;
}
.section > h2 {
  font-size: 3rem;
}
.section > p {
  font-size: 2rem;
}
.section > button {
  background-color: var(--darkerGray);
  border: none;
  border-radius: 5px;
  padding: 1rem 2rem;
  font-size: 1rem;
  cursor: pointer;
}

.section > button:hover {
  filter: brightness(1.1);
}

.section > button:active {
  filter: brightness(0.9);
}
