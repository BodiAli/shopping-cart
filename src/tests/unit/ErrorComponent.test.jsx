import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import ErrorComponent from "../../components/ErrorComponent/ErrorComponent";
import { vi } from "vitest";

vi.mock("react-router-dom", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useRouteError: vi.fn(() => new Error("Failed to fetch")),
  };
});

const router = createMemoryRouter([
  {
    path: "/",
    element: <ErrorComponent />,
  },
]);

const wrapper = () => <RouterProvider router={router} />;

describe("Error component", () => {
  test("should render correct elements", () => {
    const { container } = render(<ErrorComponent />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test("should render unexpected error occurred", () => {
    render(<ErrorComponent />, { wrapper });
    expect(screen.getByRole("heading")).toHaveTextContent(/unexpected error occurred/i);
  });

  test("should render error message", () => {
    render(<ErrorComponent />, { wrapper });
    expect(screen.getByRole("paragraph")).toHaveTextContent(/failed to fetch/i);
  });

  test("should render a reload button", () => {
    render(<ErrorComponent />, { wrapper });
    expect(screen.getByRole("button", { name: /retry/i })).toBeInTheDocument();
  });
});
