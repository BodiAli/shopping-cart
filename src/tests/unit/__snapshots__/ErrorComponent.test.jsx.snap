// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error component > should render correct elements 1`] = `
<div>
  <section
    class="_section_6dffa1"
  >
    <h2>
      Unexpected error occurred!
    </h2>
    <p>
      Failed to fetch
    </p>
    <button>
      Retry
    </button>
  </section>
</div>
`;
